### ISSUE TITLE
Add additional helpers for combining Resources/Outputs

### ISSUE THREAD

**Issue by lukehoban** [pulumi/pulumi#994]
As we have worked on fixes to explicit dependency tracking in pulumi-cloud, we have seen some patterns which could use additional helpers in the `@pulumi/pulumi` package:

First, a way to take a Resource and turn it into an Output<T> with a dependency on the resource.  

Second, a way to take a Resource and get an Output<T> with dependencies on the Resource plus additional resources.

These could be new overloads of `pulumi.output` (like `pulumi.output(myresource, [otherresource, andanothresource])`.

And/or they could be a helper on `Resource` (like `myresource.after(otherresource, andanotherresource` (or `myresource.output(...)` for more consistency with existing naming?)).

**Comment by joeduffy** [comment]
@jen20 and I ran into the need for another combinator: `toMap`.

Consider this HCL2 program:

```
output "instance_private_ip_addresses" {
    value = {
        for instance in aws_instance.example:
        instance.id -> instance.private_ip
    }
}
```

This can be written in Pulumi roughly as:

```
export let instancePrivateIpAddresses = pulumi.output(
    instances.map(i => [i.id, i.ami])).apply(kvps => new Map(kvps));
```

But it's so much nicer if we add a hypothetical `toMap` combinator:

```
export let instancePrivateIpAddresses = instances.map(i => [i.id, i.instanceIp]).toMap();
```

**Comment by lukehoban** [comment]
Do we have any concrete proposal on the table for this yet? (addressing in particular the original topic of this issue about getting an Output from a Resource).  Let's aim to propose something concrete to do here for M20.

**Comment by lukehoban** [comment]
Related to #2295 which may be (part of) the way we address this.

**Comment by lukehoban** [comment]
We landed #2295, which is a good solution to the most common case of this. We'll push any more fine grained solutions to look at in the future once we have more experience.

**Comment by CyrusNajmabadi** [comment]
Closing out for now, absent any driving scenarios that really need assistance.  If we run into cases in the future, we can open new issues.
